LibVMI doesn't compile on aarch64

Please see my steps and the output below. A git bisection was done; results at bottom.

/home/matt/libvmi/build# cmake -DCMAKE_INSTALL_PREFIX=/usr/local -DENABLE_KVM=OFF -DENABLE_FILE=OFF  ..
-- The C compiler identification is GNU 8.3.0
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29")
-- Checking for one of the modules 'glib-2.0'
-- Could NOT find LibvmiRequest (missing:  LibvmiRequest_INCLUDE_DIR)
-- Looking for xenctrl.h
-- Looking for xenctrl.h - found
-- Looking for xen/io/ring.h
-- Looking for xen/io/ring.h - found
-- Looking for sys/types.h 
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of hvmmem_access_t
-- Check size of hvmmem_access_t - failed
-- Check size of xenmem_access_t
-- Check size of xenmem_access_t - done
-- Found Xen: /usr/local/include
-- Found Xenstore: /usr/local/include
-- Found FLEX: /usr/bin/flex (found version "2.6.1")
-- Found BISON: /usr/bin/bison (found version "3.0.4")
-- Found JSON-C: /usr/lib/aarch64-linux-gnu/libjson-c.so
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
CMake Warning at doc/CMakeLists.txt:15 (message):
  Doxygen not found.  Target doc will not be available


--  ↳ Compiler     → /usr/bin/cc
--  ↳ C Flags      →  -Wall -Wextra
--  ↳ Prefix       → /usr/local
--  ↳ Build        →
-- Configuring done
-- Generating done
-- Build files have been written to: /home/matt/libvmi/build
libvmi/build# make
Scanning dependencies of target vmi
[  1%] Building C object libvmi/CMakeFiles/vmi.dir/accessors.c.o
[  2%] Building C object libvmi/CMakeFiles/vmi.dir/convenience.c.o
[  4%] Building C object libvmi/CMakeFiles/vmi.dir/core.c.o
[  5%] Building C object libvmi/CMakeFiles/vmi.dir/events.c.o
[  6%] Building C object libvmi/CMakeFiles/vmi.dir/performance.c.o
[  8%] Building C object libvmi/CMakeFiles/vmi.dir/pretty_print.c.o
[  9%] Building C object libvmi/CMakeFiles/vmi.dir/read.c.o
[ 10%] Building C object libvmi/CMakeFiles/vmi.dir/slat.c.o
[ 12%] Building C object libvmi/CMakeFiles/vmi.dir/strmatch.c.o
[ 13%] Building C object libvmi/CMakeFiles/vmi.dir/write.c.o
[ 14%] Building C object libvmi/CMakeFiles/vmi.dir/memory.c.o
[ 16%] Building C object libvmi/CMakeFiles/vmi.dir/arch/arch_interface.c.o
[ 17%] Building C object libvmi/CMakeFiles/vmi.dir/arch/intel.c.o
[ 18%] Building C object libvmi/CMakeFiles/vmi.dir/arch/amd64.c.o
[ 20%] Building C object libvmi/CMakeFiles/vmi.dir/arch/arm_aarch32.c.o
[ 21%] Building C object libvmi/CMakeFiles/vmi.dir/arch/arm_aarch64.c.o
[ 22%] Building C object libvmi/CMakeFiles/vmi.dir/driver/driver_interface.c.o
[ 24%] Building C object libvmi/CMakeFiles/vmi.dir/driver/memory_cache.c.o
/home/matt/libvmi/libvmi/driver/memory_cache.c: In function 'memory_cache_destroy':
/home/matt/libvmi/libvmi/driver/memory_cache.c:266:48: warning: cast between incompatible function types from 'void (*)(void *)' to 'void (*)(void *, void *)' [-Wcast-function-type]
         g_queue_foreach(vmi->memory_cache_lru, (GFunc)g_free, NULL);
                                                ^
[ 25%] Building C object libvmi/CMakeFiles/vmi.dir/os/os_interface.c.o
[ 27%] Building C object libvmi/CMakeFiles/vmi.dir/os/windows/core.c.o
/home/matt/libvmi/libvmi/os/windows/core.c: In function 'pe2version':
/home/matt/libvmi/libvmi/os/windows/core.c:105:13: warning: this statement may fall through [-Wimplicit-fallthrough=]
             switch (minor_os_version) {
             ^~~~~~
/home/matt/libvmi/libvmi/os/windows/core.c:113:9: note: here
         case 6:
         ^~~~
/home/matt/libvmi/libvmi/os/windows/core.c:114:13: warning: this statement may fall through [-Wimplicit-fallthrough=]
             switch (minor_os_version) {
             ^~~~~~
/home/matt/libvmi/libvmi/os/windows/core.c:122:9: note: here
         case 10:
         ^~~~
[ 28%] Building C object libvmi/CMakeFiles/vmi.dir/os/windows/kdbg.c.o
[ 29%] Building C object libvmi/CMakeFiles/vmi.dir/os/windows/memory.c.o
[ 31%] Building C object libvmi/CMakeFiles/vmi.dir/os/windows/peparse.c.o
[ 32%] Building C object libvmi/CMakeFiles/vmi.dir/os/windows/process.c.o
[ 33%] Building C object libvmi/CMakeFiles/vmi.dir/os/windows/unicode.c.o
[ 33%] Built target vmi
[ 35%] Generating parser.c
[ 36%] Generating scanner.c
Scanning dependencies of target vmi_shared
[ 37%] Building C object libvmi/CMakeFiles/vmi_shared.dir/scanner.c.o
/home/matt/libvmi/build/libvmi/scanner.c: In function 'yy_get_next_buffer':
/home/matt/libvmi/build/libvmi/scanner.c:736:18: warning: comparison of integer expressions of different signedness: 'size_t' {aka 'long unsigned int'} and 'int' [-Wsign-compare]
   for ( n = 0; n < max_size && \
                  ^
/home/matt/libvmi/build/libvmi/scanner.c:1338:3: note: in expansion of macro 'YY_INPUT'
  YY_INPUT( (&YY_CURRENT_BUFFER_LVALUE->yy_ch_buf[number_to_move]),
   ^~~~~~~~
[ 39%] Building C object libvmi/CMakeFiles/vmi_shared.dir/parser.c.o
[ 40%] Building C object libvmi/CMakeFiles/vmi_shared.dir/driver/xen/altp2m.c.o
[ 41%] Building C object libvmi/CMakeFiles/vmi_shared.dir/driver/xen/xen.c.o
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_discover_pv_type':
/home/matt/libvmi/libvmi/driver/xen/xen.c:440:1: warning: label '_bail' defined but not used [-Wunused-label]
 _bail:
 ^~~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:394:20: warning: unused parameter 'vmi' [-Wunused-parameter]
     vmi_instance_t vmi)
     ~~~~~~~~~~~~~~~^~~
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_init':
/home/matt/libvmi/libvmi/driver/xen/xen.c:572:37: warning: cast between incompatible function types from 'int (*)(const void *, const void *)' to 'gint (*)(const void *, const void *, void *)' {aka 'int (*)(const void *, const void *, void *)'} [-Wcast-function-type]
     xen->domains = g_tree_new_full ((GCompareDataFunc)domains_compare, NULL, key_destroy_func, value_destroy_func);
                                     ^
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_get_xsave_info':
/home/matt/libvmi/libvmi/driver/xen/xen.c:879:31: error: 'CPU_XSAVE_CODE' undeclared (first use in this function); did you mean 'HVM_SAVE_CODE'?
         if (desc->typecode == CPU_XSAVE_CODE && desc->instance == vcpu) {
                               ^~~~~~~~~~~~~~
                               HVM_SAVE_CODE
/home/matt/libvmi/libvmi/driver/xen/xen.c:879:31: note: each undeclared identifier is reported only once for each function it appears in
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_get_vcpumtrr':
/home/matt/libvmi/libvmi/driver/xen/xen.c:2543:20: warning: unused parameter 'vmi' [-Wunused-paramete ]
     vmi_instance_t vmi,
     ~~~~~~~~~~~~~~~^~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:2544:18: warning: unused parameter 'hwMtrr' [-Wunused-parameter]
     mtrr_regs_t *hwMtrr,
     ~~~~~~~~~~~~~^~~~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:2545:19: warning: unused parameter 'vcpu' [-Wunused-parameter]
     unsigned long vcpu)
     ~~~~~~~~~~~~~~^~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_get_vcpuregs':
/home/matt/libvmi/libvmi/driver/xen/xen.c:2580:20: warning: unused parameter 'vmi' [-Wunused-paramete ]
     vmi_instance_t vmi,
     ~~~~~~~~~~~~~~~^~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:2581:18: warning: unused parameter 'regs' [-Wunused-parameter]
     registers_t *regs,
     ~~~~~~~~~~~~~^~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:2582:19: warning: unused parameter 'vcpu' [-Wunused-parameter]
     unsigned long vcpu)
     ~~~~~~~~~~~~~~^~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_set_vcpuregs':
/home/matt/libvmi/libvmi/driver/xen/xen.c:2617:20: warning: unused parameter 'vmi' [-Wunused-paramete ]
     vmi_instance_t vmi,
     ~~~~~~~~~~~~~~~^~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:2618:18: warning: unused parameter 'regs' [-Wunused-parameter]
     registers_t *regs,
     ~~~~~~~~~~~~~^~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c:2619:19: warning: unused parameter 'vcpu' [-Wunused-parameter]
     unsigned long vcpu)
     ~~~~~~~~~~~~~~^~~~
/home/matt/libvmi/libvmi/driver/xen/xen.c: In function 'xen_is_pv':
/home/matt/libvmi/libvmi/driver/xen/xen.c:2651:20: warning: unused parameter 'vmi' [-Wunused-paramete ]
     vmi_instance_t vmi)   
     ~~~~~~~~~~~~~~~^~~
libvmi/CMakeFiles/vmi_shared.dir/build.make:142: recipe for target 'libvmi/CMakeFiles/vmi_shared.dir/driver/xen/xen.c.o' failed
make[2]: *** [libvmi/CMakeFiles/vmi_shared.dir/driver/xen/xen.c.o] Error 1
CMakeFiles/Makefile2:158: recipe for target 'libvmi/CMakeFiles/vmi_shared.dir/all' failed
make[1]: *** [libvmi/CMakeFiles/vmi_shared.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
# uname -a
Linux xilinx-zcu102-zu9-es2-rev1_0-2018.3 4.14.0-xilinx-v2018.3 #1 SMP Thu Dec 6 10:01:26 UTC 2018 aarch64 GNU/Linux
# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/aarch64-linux-gnu/8/lto-wrapper
Target: aarch64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 8.3.0-6' --with-bugurl=file:///usr/share/doc/gcc-8/README.Bugs --enable-languages=c,ada,c++,go,d,fortran,objc,obj-c++ --prefix=/usr --with-gcc-major-version-only --program-suffix=-8 --program-prefix=aarch64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --enable-bootstrap --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-libquadmath --disable-libquadmath-support --enable-plugin --enable-default-pie --with-system-zlib --disable-libphobos --enable-multiarch --enable-fix-cortex-a53-843419 --disable-werror --enable-checking=release --build=aarch64-linux-gnu --host=aarch64-linux-gnu --target=aarch64-linux-gnu
Thread model: posix
gcc version 8.3.0 (Debian 8.3.0-6) 
# git reflog
3454947 HEAD@{0}: checkout: moving from master to master


A git bisection shows this:
2b67e0547bbb0cfa008c7cb6894fe25ec43a2e4b is the first bad commit
commit 2b67e0547bbb0cfa008c7cb6894fe25ec43a2e4b
Author: Isaila Alexandru <47214621+aisaila@users.noreply.github.com>
Date:   Thu Apr 18 01:22:48 2019 +0300

    Add xen_get_xsave_info (#733)
    
    This patch adds a new function to extract the cpu xsave info.
    
    Signed-off-by: Alexandru Isaila <aisaila@bitdefender.com>

:040000 040000 33e0f324a7a84ce34637f3b9bc93870f3f24c0e0 e2d6bb7d32f4887226e59778678fcef736f57a56 M      libvmi

--- a/configure.ac
+++ b/configure.ac
@@ -129,7 +129,6 @@ AC_ARG_ENABLE([vmifs],
          [Disable building VMIFS tool: maps memory to a file through FUSE @<:@no@:>@])],
       [enable_vmifs=$enableval],
       [enable_vmifs=yes])
-AM_CONDITIONAL([VMIFS], [test x"$enable_vmifs" = xyes])
 
 AC_ARG_ENABLE([address_cache],
       [AS_HELP_STRING([--disable-address-cache],
@@ -150,7 +149,6 @@ AC_ARG_ENABLE([rekall_profiles],
          [Disable using Rekall's JSON profiles @<:@no@:>@])],
       [rekall="$enableval"],
       [rekall="yes"])
-AM_CONDITIONAL([ENABLE_REKALL_PROFILES], [test x"$rekall" = "xyes"])
 
 AC_ARG_ENABLE([config-file],
       [AS_HELP_STRING([--disable-config-file],
@@ -287,6 +285,7 @@ vmifs_space='       '
         AC_SUBST([FUSE_LIBS])
     [fi]
 [fi]
+AM_CONDITIONAL([VMIFS], [test x"$enable_vmifs" = xyes])
 
 [if test "$configfile" = "yes"]
 [then]
@@ -322,14 +321,14 @@ missing='no'
     AC_CHECK_LIB(json-c, json_object_get_int64,, [missing="yes"])
     [if test x"$missing" = "xno"]
     [then]
-            AC_CHECK_HEADERS([json-c/json.h])
-            AC_DEFINE([REKALL_PROFILES], [1], [Defined to 1 when working JSON-C library was found to parse Rekall profiles.])
+        AC_CHECK_HEADERS([json-c/json.h])
+        AC_DEFINE([REKALL_PROFILES], [1], [Defined to 1 when working JSON-C library was found to parse Rekall profiles.])
     [else]
         rekall='Disabled, missing libjson-c-dev.'
         [echo "No working JSON-C library found (libjson-c-dev), Rekall system profiles won't be supported."]
     [fi]
 [fi]
-AM_CONDITIONAL([REKALL_PROFILES], [test x"$rekall" = xyes])
+AM_CONDITIONAL([ENABLE_REKALL_PROFILES], [test x"$rekall" = "xyes"])
 
 [if test x"$enable_address_cache" = "xyes"]
 [then]

--- a/libvmi/driver/xen/xen.c
+++ b/libvmi/driver/xen/xen.c
@@ -812,7 +812,8 @@ xen_get_tsc_info(
     return VMI_SUCCESS;
 }
 
-void
+#if defined(I386) || defined(X86_64)
+static void
 xen_get_xsave_info_46(
     struct hvm_hw_cpu_xsave_46 *info,
     xsave_area_t *xsave_info)
@@ -822,7 +823,7 @@ xen_get_xsave_info_46(
     xsave_info->xstate_bv = info->save_area.xsave_hdr.xstate_bv;
 }
 
-void
+static void
 xen_get_xsave_info_412(
     struct hvm_hw_cpu_xsave_412 *info,
     xsave_area_t *xsave_info)
@@ -832,8 +833,8 @@ xen_get_xsave_info_412(
     xsave_info->xstate_bv = info->save_area.xsave_hdr.xstate_bv;
 }
 
-status_t
-xen_get_xsave_info(
+static status_t
+xen_get_xsave_info_hvm(
     vmi_instance_t vmi,
     unsigned long vcpu,
     xsave_area_t *xsave_info)
@@ -906,8 +907,6 @@ _bail:
 
     return ret;
 }
-
-#if defined(I386) || defined(X86_64)
 
 static status_t
 xen_get_vcpumtrr_hvm(
@@ -2535,6 +2534,20 @@ xen_get_vcpumtrr(
 #if defined(I386) || defined (X86_64)
     if (vmi->vm_type == HVM)
         return xen_get_vcpumtrr_hvm(vmi, hwMtrr, vcpu);
+#endif
+
+    return VMI_FAILURE;
+}
+
+status_t
+xen_get_xsave_info(
+    vmi_instance_t vmi,
+    unsigned long vcpu,
+    xsave_area_t *xsave_info)
+{
+#if defined(I386) || defined (X86_64)
+    if (vmi->vm_type == HVM)
+        return xen_get_xsave_info_hvm(vmi, vcpu, xsave_info);
 #endif
 
     return VMI_FAILURE;

--- a/libvmi/read.c
+++ b/libvmi/read.c
@@ -48,7 +48,7 @@ vmi_mmap_guest(
     addr_t paddr;
     size_t buf_offset = 0;
     unsigned long *pfns = NULL;
-    unsigned int pfn_ndx = 0;
+    unsigned int pfn_ndx = 0, i;
 
     switch (ctx->translate_mechanism) {
         case VMI_TM_KERNEL_SYMBOL:
@@ -94,7 +94,7 @@ vmi_mmap_guest(
 
     pfns = calloc(num_pages, sizeof(unsigned long));
 
-    for (unsigned int i = 0; i < num_pages; i++) {
+    for (i = 0; i < num_pages; i++) {
         if (VMI_SUCCESS == vmi_pagetable_lookup_cache(vmi, dtb, vaddr + buf_offset, &paddr)) {
             pfns[pfn_ndx] = paddr >> vmi->page_shift;
             // store relative offsets to the appropriate pages
@@ -119,7 +119,7 @@ vmi_mmap_guest(
         goto done;
     }
 
-    for (unsigned int i = 0; i < num_pages; i++) {
+    for (i = 0; i < num_pages; i++) {
         if (access_ptrs[i] != (void *)-1) {
             // add buffer base pointer to the relative offsets since now we know its value
             access_ptrs[i] += (addr_t)base_ptr;

